numpyro/infer/svi.py

```diff
@@ -137,6 +135,45 @@ def update(self, svi_state, *args, **kwargs):
         loss_val, optim_state = self.optim.eval_and_update(loss_fn, svi_state.optim_state)
         return SVIState(optim_state, rng_key), loss_val
 
+    def run(self, rng_key, num_steps, *args, progress_bar=True, **kwargs):
+        """
+        (EXPERIMENTAL INTERFACE) Run SVI with `num_steps` iterations, then return
+        the optimized parameters and the stacked losses at every step. If `num_steps`
+        is large, setting `progress_bar=False` can make the run faster.
+
+        .. note:: For a complex training process (e.g. the one requires early stopping,
+            epoch training, varying args/kwargs,...), we recommend to use the more
+            flexible methods :meth:`init`, :meth:`update`, :meth:`evaluate` to
+            customize your training procedure.
+
+        :param jax.random.PRNGKey rng_key: random number generator seed.
+        :param int num_steps: the number of optimization steps.
+        :param args: arguments to the model / guide
+        :param bool progress_bar: Whether to enable progress bar updates. Defaults to
+            ``True``.
+        :param kwargs: keyword arguments to the model / guide
+        :return: a tuple of `(params, losses)` where `params` holds the optimized values
+            at :class:`numpyro.param` sites, and `losses` is the collected loss
+            during the process.
+        """
+        def body_fn(svi_state, carry):
+            svi_state, loss = self.update(svi_state, *args, **kwargs)
+            return svi_state, loss
+
+        svi_state = self.init(rng_key, *args, **kwargs)
+        if progress_bar:
+            losses = []
+            with tqdm.trange(num_steps) as t:
+                for i in t:
+                    svi_state, loss = jit(body_fn)(svi_state, None)
+                    t.set_postfix_str("Loss: {:.4e}".format(loss), refresh=False)
+                    losses.append(loss)
+            losses = jnp.stack(losses)
+        else:
+            svi_state, losses = lax.scan(body_fn, svi_state, None, length=num_steps)
+
+        return self.get_params(svi_state), losses
```

Comment:
Do we foresee other things that we might want to return in the future (I can't think of anything at the moment)? If so, it will be better to use a named tuple so that we don't need to introduce a backwards incompatible change in the future.

Reply:
Makes sense to me. I'll make the change.

Reply:
One thing we might want to return is the last `svi_state`, which is used for evaluation. But I am not sure how useful it is. For evaluation, we can construct `svi_state` back from `params` by using `SVIState(self.optim.init(params), rng_key)`. To me, it is better to have a method `get_state(self, params, rng_key)` if `svi_state` is needed.

The current return values seem good to me.

Reply:
I see - yes, we don't need to add anything else right now. I was only asking since if we wanted to have such an option for the future, we could build that flexibility in right now.